lib/auth0/mixins/validation.rb

```diff
@@ -20,6 +28,118 @@ def validate_permissions_array(permissions)
+          rescue JWT::VerificationError
+            raise Auth0::InvalidIdToken, 'Invalid ID token signature'
+          rescue JWT::IncorrectAlgorithm
+            alg = decoded_id_token[:header]['alg']
+            raise Auth0::InvalidIdToken, "Signature algorithm of \"#{alg}\" is not supported. Expected the ID token to be signed with \"#{algorithm_name}\""
+          rescue JWT::DecodeError
+            kid = decoded_id_token[:header]['kid']
+            raise Auth0::InvalidIdToken, "Could not find a public key for Key ID (kid) \"#{kid}\""
+          end
+        end
+
+        def validate_claims(decoded_id_token); end
+      end
+
+      module Algorithm
+        # Represents the HS256 algorithm, which rely on shared secrets.
+        # @see https://auth0.com/docs/tokens/concepts/signing-algorithms
+        class HS256
+          class << self
+            private :new
+
+            # Create a new instance passing the shared secret.
+            # @param secret [string] The HMAC shared secret.
+            # @return [object] A new instance.
+            def secret(secret)
+              new secret
+            end
+          end
+
+          attr_accessor :secret
+
+          def initialize(secret)
+            raise Auth0::InvalidParameter, 'Must supply a valid secret' if secret.to_s.empty?
+
+            @secret = secret
+          end
+        end
+
+        # Represents the RS256 algorithm, which rely on public key certificates.
+        # @see https://auth0.com/docs/tokens/concepts/signing-algorithms
+        class RS256
+          include Auth0::Mixins::HTTPProxy
+
+          @@cache = Zache.new.freeze
+
+          class << self
+            private :new
+
+            # Create a new instance passing the shared secret.
+            # @param secret [string] The url where the JWK set is located.
+            # @param lifetime [integer] The lifetime of the JWK set in-memory cache in seconds.
+            #   Must be a non-negative value. Defaults to 600 seconds (10 minutes).
+            # @return [object] A new instance.
+            def jwks_url(url, lifetime: 10 * 60)
+              new url, lifetime
+            end
+          end
+
+          def initialize(jwks_url, lifetime)
+            raise Auth0::InvalidParameter, 'Must supply a valid jwks_url' if jwks_url.to_s.empty?
+            raise Auth0::InvalidParameter, 'Must supply a valid lifetime' unless lifetime.is_a?(Integer) && lifetime >= 0
+
+            @lifetime = lifetime
+            @jwks_url = jwks_url
+          end
+
+          # Fetches the JWK set from the in-memory cache, or from the url as a fallback.
+          # @return [hash] A JWK set.
+          def jwks
+            previous_value = @@cache.last(:jwks)[:value] unless @@cache.last(:jwks).nil?
+
+            @@cache.get(:jwks, lifetime: @lifetime, dirty: true) do
+              result = get(@jwks_url)
+              result.is_a?(Hash) && result.key?('keys') ? result : previous_value
```

Comment:
I'd expect an explicit set but maybe the library is doing that for you. Anyway, will the cache be updated with this result for the next call?

Reply:
Ruby has implicit returns, so that line (the last one in the block) will return the jwks obtained from the network call or the last known value otherwise.

Reply:
The [return value](https://github.com/yegor256/zache/blob/master/lib/zache.rb#L187) of that block is what ends up in the cache.